TypeScript error in adapt/task.ts

## Error outputs

```
src/ts/adapt/task.ts(120,19): error TS2345: Argument of type 'Timer | TimerImpl' is not assignable to parameter of type 'Timer'.
  Type 'TimerImpl' is not assignable to type 'Timer'.
    Types of property 'setTimeout' are incompatible.
      Type '(fn: () => void, delay: number) => Timer' is not assignable to type '(fn: () => void, delay: number) => number'.
        Type 'Timer' is not assignable to type 'number'.
src/ts/adapt/task.ts(179,3): error TS2416: Property 'setTimeout' in type 'TimerImpl' is not assignable to the same property in base type 'Timer'.
  Type '(fn: () => void, delay: number) => Timer' is not assignable to type '(fn: () => void, delay: number) => number'.
    Type 'Timer' is not assignable to type 'number'.
src/ts/adapt/task.ts(279,52): error TS2314: Generic type 'Continuation<T>' requires 1 type argument(s).
src/ts/adapt/task.ts(353,33): error TS2314: Generic type 'Continuation<T>' requires 1 type argument(s).
src/ts/adapt/task.ts(596,24): error TS2345: Argument of type 'T1' is not assignable to parameter of type 'T'.
src/ts/adapt/task.ts(599,7): error TS2322: Type 'Result<T>' is not assignable to type 'Result<T1>'.
  Type 'T' is not assignable to type 'T1'.
src/ts/adapt/task.ts(804,28): error TS2314: Generic type 'LoopBodyFrame<T>' requires 1 type argument(s).
src/ts/adapt/task.ts(841,17): error TS2345: Argument of type 'true' is not assignable to parameter of type 'T'.
src/ts/adapt/task.ts(845,17): error TS2345: Argument of type 'false' is not assignable to parameter of type 'T'.
```

--- a/src/ts/adapt/expr.ts
+++ b/src/ts/adapt/expr.ts
@@ -66,15 +66,16 @@ export const clonePreferences = (pref: Preferences): Preferences => ({
 
 export const defaultPreferencesInstance = defaultPreferences();
 
+interface Pending {};
+type Special = Pending;
+
 /**
  * Special marker value that indicates that the expression result is being
  * calculated.
- * @enum {!Object}
  */
 export const Special = {
-  PENDING: {}
+  PENDING: {} as Pending,
 };
-type Special = typeof Special;
 
 type Result = string|number|boolean|undefined;
 
@@ -588,7 +589,7 @@ export class Prefix extends Val {
     if (val === this.val) {
       return this;
     }
-    const r = new this.constructor(this.scope, val);
+    const r = new Prefix(this.scope, val);
     return r;
   }
 }
@@ -653,7 +654,7 @@ export class Infix extends Val {
     if (lhs === this.lhs && rhs === this.rhs) {
       return this;
     }
-    const r = new this.constructor(this.scope, lhs, rhs);
+    const r = new Infix(this.scope, lhs, rhs);
     return r;
   }
 }
@@ -1035,6 +1036,10 @@ export class Named extends Val {
  * Named value.
  */
 export class MediaName extends Val {
+  // FIXME: This property is added to reduce TypeScript error on `dependCore`
+  // but it is never initialized. Is it really correct code?
+  value: any;
+
   constructor(scope: LexicalScope, public not: boolean, public name: string) {
     super(scope);
   }

--- a/src/ts/adapt/task.ts
+++ b/src/ts/adapt/task.ts
@@ -176,7 +176,12 @@ export class TimerImpl implements Timer {
   /**
    * @override
    */
-  setTimeout(fn: () => void, delay: number) {return setTimeout(fn, delay);}
+  setTimeout(fn: () => void, delay: number) {
+    // HACK: casting to unknown type to prevent TypeScript error
+    // [TS2352] Conversion of type 'Timer' to type 'number' may be a mistake because neither type sufficiently overlaps with the other.
+    const timer: unknown = this.setTimeout(fn, delay);
+    return timer as number;
+  }
 
   /**
    * @override
@@ -276,7 +281,7 @@ export class Scheduler {
       let now = this.timer.currentTime();
       this.sliceOverTime = now + this.slice;
       while (this.queue.length()) {
-        const continuation = (this.queue.peek() as Continuation);
+        const continuation = (this.queue.peek() as Continuation<any>);
         if (continuation.scheduledTime > now) {
           break;
         }
@@ -350,7 +355,7 @@ export class Continuation<T> implements base.Comparable {
    */
   compare(otherComp: base.Comparable): number {
     // earlier wins
-    const other = (otherComp as Continuation);
+    const other = (otherComp as Continuation<any>);
     return other.scheduledTime - this.scheduledTime || other.order - this.order;
   }
 
@@ -370,7 +375,7 @@ export class Continuation<T> implements base.Comparable {
     this.task.scheduler.schedule(this, opt_delay);
   }
 
-  private resumeInternal(): boolean {
+  resumeInternal(): boolean {
     const task = this.task;
     this.task = null;
     if (task && task.continuation == this) {
@@ -584,19 +589,19 @@ export class ResultImpl<T> implements Result<T> {
   /**
    * @override
    */
-  thenAsync<T1>(callback: (p1: T) => Result<T1>) : Result<T1>{
+  thenAsync<T1>(callback: (p1: T) => Result<T1>): Result<T1>{
     if (this.isPending()) {
       // thenAsync is special, do the trick with the context
-      const frame = new Frame<T>(
+      const frame = new Frame<T | T1>(
           this.frame.task, this.frame.parent, 'AsyncResult.thenAsync');
       frame.state = FrameState.ACTIVE;
-      this.frame.parent = frame;
+      this.frame.parent = frame as Frame<T>;
       this.frame.then((res1) => {
         callback(res1).then((res2) => {
           frame.finish(res2);
         });
       });
-      return frame.result();
+      return frame.result() as Result<T1>;
     } else {
       return callback(this.frame.res);
     }
@@ -832,8 +837,8 @@ export class Frame<T> {
   }
 }
 
-export class LoopBodyFrame<T> extends Frame<T> {
-  constructor(task: Task, parent: Frame<T>) {
+export class LoopBodyFrame extends Frame<boolean> {
+  constructor(task: Task, parent: Frame<boolean>) {
     super(task, parent, 'loop');
   }
 

--- a/src/ts/adapt/taskutil.ts
+++ b/src/ts/adapt/taskutil.ts
@@ -98,14 +98,14 @@ export class Fetcher<T> {
 /**
  * Wait for all Fetcher objects in the array to arrive
  */
-export const waitForFetchers = (fetchers: Fetcher[]): task.Result<boolean> => {
+export const waitForFetchers = <T>(fetchers: Fetcher<T>[]): task.Result<boolean> => {
   if (fetchers.length == 0) {
     return task.newResult(true);
   }
   if (fetchers.length == 1) {
     return fetchers[0].get().thenReturn(true);
   }
-  const frame = task.newFrame('waitForFetches');
+  const frame = task.newFrame<boolean>('waitForFetches');
   let i = 0;
   frame
       .loop(() => {

